MSC2398: proposal to allow mxc:// in the "a" tag within messages

eras opened the pull request:
Currently message events (m.room.message) can contain formatted data in its formatted_body field, if the format of the message is org.matrix.custom.html. This is a proposal to allow the `mxc://` schema to appear also for `a` tags in addition to just `img` tags currently permitted (in fact, required).

turt2live commented on specification/modules/instant_messaging.rst:
```diff
@@ -86,7 +86,8 @@ for the tag.
 :``a``:
   ``name``, ``target``, ``href`` (provided the value is not relative and has a scheme
-  matching one of: ``https``, ``http``, ``ftp``, ``mailto``, ``magnet``)
+  matching one of: ``https``, ``http``, ``ftp``, ``mailto``, ``magnet``, or ``mxc``
```
It's very annoying, but this change will need to happen after the proposal goes through the first few steps of the process. Particularly it'll need implementation proof and a few other supporting pieces to be properly merged. Generally speaking, because the post-FCP process is annoying, I don't mind personally doing the spec work to get it written up.

eras replied:
Removed changes to `instant_messaging.rst`

turt2live commented:
@eras also slightly annoying, but the proposal needs to be signed off so we can merge it: https://github.com/matrix-org/matrix-doc/blob/master/CONTRIBUTING.rst#sign-off

eras commented:
Updated.

Changes:
- Added note about alternatives per @KitsunRal 's comment 
- Removed changes to `instant_messaging.rst` per @turt2live 's comment
- Added a Signed Off -line to the commit message per @turt2live 's comment

eras commented:
So it came to me that an implementation should be careful when implementing this.

In particular, when a web-based client opens a HTML document from the same domain as the Matrix API, the document can probably use JavaScript to make requests with the user's credentials.

ShadowJonathan commented:
@eras i believe those thoughts would be best served under the "Security Considerations" header in the proposal :+1:

turt2live commented on proposals/2398-allow-mxc-within-message-links.md:
```diff
@@ -0,0 +1,55 @@
+
+## Proposal
+
+The spec for `a` tags says:
```
```suggestion
The spec for `a` tags [says](https://spec.matrix.org/v1.8/client-server-api/#mroommessage-msgtypes):
```

turt2live reviewed (approved):
non-blocking concerns

turt2live commented:
sorry this got stuck :(

@mscbot fcp merge

dbkr commented on proposals/2398-allow-mxc-within-message-links.md:
```diff
@@ -0,0 +1,55 @@
+# MSC2398: Proposal to allow mxc:// in the `a` tag within messages
+
+Currently message events (m.room.message) can contain formatted data
+in its formatted_body field, if the format of the message is
```
```suggestion
in their formatted_body field, if the format of the message is
```

dbkr commented on proposals/2398-allow-mxc-within-message-links.md:
```diff
@@ -0,0 +1,55 @@
+
+Currently message events (m.room.message) can contain formatted data
+in its formatted_body field, if the format of the message is
+org.matrix.custom.html. In this subset of HTML `img` tags are permitted
```
```suggestion
org.matrix.custom.html. In this subset of HTML, `img` tags are permitted
```

dbkr commented on proposals/2398-allow-mxc-within-message-links.md:
```diff
@@ -0,0 +1,55 @@
+in its formatted_body field, if the format of the message is
+org.matrix.custom.html. In this subset of HTML `img` tags are permitted
+if they refer to media repository contents with the mxc schema and
+clients are able to resolve these URIs to (ie.) HTTPs to retrieve the
```
```suggestion
clients are able to resolve these URIs to (eg.) HTTP(s) URLs to retrieve the
```

Sorry if these feel nitpicky, I'm just having to re-read a couple of times to work out what the MSC is saying so attempting to clarify.

dbkr commented on proposals/2398-allow-mxc-within-message-links.md:
```diff
@@ -0,0 +1,55 @@
+-`a`:
+  `name`, `target`, `href` (provided the value is not relative and has a scheme
+  matching one of: `https`, `http`, `ftp`, `mailto`, `magnet`, or `mxc`
+  for `Matrix Content (MXC) URI`)
```
Much as it's obvious, presumably the spec would also need to explicitly say that the client must transform the mxc link into an appropriate link the browser will understand (ie. what you talk about above with img tags).

turt2live replied:
This feels like something which is best served by a note rather than requirement. Web clients will probably need to convert it, but other clients might be able to handle a custom scheme for network requests. 

I'd suggest that this *not* be a blocking concern for FCP, as the spec text can provide guidance on how to handle it.

KitsuneRal replied:
I would really prefer the MSC say explicitly that clients should manage this scheme in one way or another - ultimately a custom handler will have to translate the `mxc` URL into an HTTP(S) request anyway.

KitsuneRal replied:
Oh wait but isn't "Clients would be able to handle the mxc schema here in the exact same way they handle `img` tags" above exactly this description, actually? I personally think that is enough.

anoadragon453 replied:
What do you think @dbkr?

dbkr replied:
I guess it just seemed odd to me that an explicit change to the spec text was given but then additional behaviour specified in the MSC. I suppose it's all there though so it doesn't really matter.

dbkr reviewed (commented):
Going to raise a concern for adding the bit about transforming the mxc link just so there's a clear action point, but just to say that in spite of this and the grammar nitpicks,  this is a superbly written and simple, but thorough, MSC, and in general seems like a great change. Thank you!

dbkr commented:
@mscbot concern add explicit text about transforming mxc link to https

dbkr commented:
@mscbot resolve add explicit text about transforming mxc link to https

ara4n commented on proposals/2398-allow-mxc-within-message-links.md:
```diff
@@ -0,0 +1,55 @@
+retrieve resources linked this way. Some clients may already
+accidentally support this, though, due to consistent handling of URIs.
+
+## Security considerations ##
```
It's probably worth spelling out explicitly here that the security risks of linking directly to content in the repository is not just HTML, but any format which can execute code and potentially XSS needs to be mitigated by a valid CSP - e.g. JS, PDF, SWF, SVG.  This isn't anything new, but it's still a security consideration.  I can't think of a world where a HTTPS url derived from an MXC URL wouldn't be subject to CSP when a plain HTTPS url to the same content would, but still worth spelling it out at least in the MSC, if not the final spec.

turt2live commented on proposals/2398-allow-mxc-within-message-links.md:
```diff
@@ -0,0 +1,55 @@
+# MSC2398: Proposal to allow mxc:// in the `a` tag within messages
```
The media ID would be attributed to a single event somewhere, and clients would be translating `mxc://` to `https://` for the user. This then behaves no differently than if a user were to get linked to the download endpoint directly - authentication would be required, and if that user can't see the original event then the link is broken for that user.

erikjohnston replied:
> This then behaves no differently than if a user were to get linked to the download endpoint directly - authentication would be required, and if that user can't see the original event then the link is broken for that user.

But getting linked directly to the download endpoint won't work, as there is no way for a user to authenticate AIUI? I thought for media web clients would potentially need to download the media in service workers or whatever to make the authentication work, in a similar manner as today to make encrypted attachments work. Is that method viable here?

turt2live replied:
A service worker would be required, which should be workable. The client can essentially iframe it and append the appropriate headers. This then ties into Matthew's comment about security considerations: https://github.com/matrix-org/matrix-spec-proposals/pull/2398#discussion_r1450850788

I'm tempted to suggest a blocking concern on this as it feels valuable to test/implement, particularly while MSC3916 and friends are being experimented with. It's akin to custom emoji/stickers where I think we want to avoid making the problem worse short-term if we can avoid it.

turt2live commented:
(FCP is not complete because there are outstanding concerns - see https://github.com/matrix-org/mscbot-python/issues/3 )

turt2live commented:
The outstanding concern on this does not appear easily resolved - I'm cancelling FCP or returning this MSC back to its pre-FCP state (whichever the bot lets me do) pending further work in the area of media linking. It feels like changes may be required that would warrant a new FCP call anyways.

@mscbot fcp cancel
